### ISSUE TITLE
Random ←[?1h is showing

### ISSUE THREAD

**Issue by shankiflang** [zsh-users/zsh-autosuggestions#614]
### Describe the bug
When I run any command I get characters like ←[?1h that go everywhere every time I type a character it doesn't know (for example a command that is not in the history)

### Screenshots
![Capture](https://user-images.githubusercontent.com/32774804/120994865-33e2a000-c785-11eb-9f4f-0d7ff736a510.PNG)

### Desktop
 - Windows 10
 - Zsh version: 5.8

**Comment by zhangchenxu** [comment]
Same to me!

Desktop
- Win10 64bit;
- Zsh version: zsh 5.8 (x86_64-pc-msys)

**Comment by kevindqc** [comment]
As a workaround, checking out the v0.6.4 tag in my `\.oh-my-zsh\custom\plugins\zsh-autosuggestions` folder fixes the issue

**Comment by zhangchenxu** [comment]
Thx, but not work for me

**Comment by zhangchenxu** [comment]
> As a workaround, checking out the v0.6.4 tag in my `\.oh-my-zsh\custom\plugins\zsh-autosuggestions` folder fixes the issue


Thx, but not work for me

**Comment by sagspot** [comment]
Am also facing the same. I don't undustand @kevindqc 's solution. Could you please explain to a newbie ☺️

**Comment by codetopper** [comment]
> Am also facing the same. I don't undustand @kevindqc 's solution. Could you please explain to a newbie ☺️

same

**Comment by kevindqc** [comment]
You can use this:
```
cd ${ZSH_CUSTOM:-~/.oh-my-zsh/custom}/plugins/zsh-autosuggestions
git checkout v0.6.4-branch
```

**Comment by sagspot** [comment]
Thank you @kevindqc. That fixed my issue.

**Comment by ericfreese** [comment]
> As a workaround, checking out the v0.6.4 tag ... fixes the issue

I don't know what might be causing this, but a big thing that changed between v0.6.4 and v0.7.0 was enabling of async mode by default. Do you see this problem in v0.7.0 with async mode disabled (`unset ZSH_AUTOSUGGEST_USE_ASYNC`)?

Does the problem occur if you run `zsh -df` (to not load any custom zsh dotfiles) and source zsh-autosuggestions manually?

> When I run any command

Are you saying the characters only appear when you press enter to run the command? Or do the characters appear before you've pressed enter. What exact action makes the characters appear?

**Comment by dbachran** [comment]
You are right, this is a minimal `~/.zshrc` to trigger this behaviour (need to also load/overwrite the predefined ZLE widgets):

```
function zle-line-init() { echoti smkx }
function zle-line-finish() { echoti rmkx }
zle -N zle-line-init
zle -N zle-line-finish
source ~/.oh-my-zsh/custom/plugins/zsh-autosuggestions/zsh-autosuggestions.zsh
```

This is the output of `infocmp -1 | grep '[rs]mkx'`:
```
        rmkx=\E[?1l\E>,
        smkx=\E[?1h\E=,
```

After adding a `sleep 4` as indicated, I get the following output:
```
Ted% a←[?1h←[?1ab←[?1h←[?1hc←[?1h
zsh: command not found: abc
Ted% abcd←[?1h←[?1h
zsh: command not found: abcd
Ted%
```

Note and further observations:
1. I type `a` and wait 4 seconds until the first `←[?1h` appears
2. When typing `b` an intial `←[?1h` immediately appears before `b`, and the second `←[?1h` appears 4 seconds later
3. Same for `c`
4. Hitting enter also adds another `←[?1h` before processing the (unknown) command, no further `←[?1h` appears afterwards
5. Typing `abc` does not produce any characters (this is known already now)
6. Typing a `d` produces another `←[?1h` after 4 seconds
7. Hitting enter adds another `←[?1h` again before processing the (unknown) command

**tl;dr**
Garbled characters are added before (immediately) **and** after (after 4 seconds) the character.

**Comment by dbachran** [comment]
Interestingly, when adding an `echo` statement within the widget function declarations (see patch below), I do not get any garbled characters anymore, but the `echo` statement instead. See output below.

Patch:
```
Ted% git diff
diff --git a/zsh-autosuggestions.zsh b/zsh-autosuggestions.zsh
index b19cac7..f214ef8 100644
--- a/zsh-autosuggestions.zsh
+++ b/zsh-autosuggestions.zsh
@@ -472,6 +472,7 @@ _zsh_autosuggest_partial_accept() {
        local action
        for action in $_ZSH_AUTOSUGGEST_BUILTIN_ACTIONS modify partial_accept; do
                eval "_zsh_autosuggest_widget_$action() {
+                       echo Z:$action
                        local -i retval

                        _zsh_autosuggest_highlight_reset
@@ -791,6 +792,8 @@ _zsh_autosuggest_async_request() {
                # Tell parent process our pid
                echo $sysparams[pid]

+               sleep 4
+
                # Fetch and print the suggestion
                local suggestion
                _zsh_autosuggest_fetch_suggestion "$1"
```

Output:
```
Ted% Z:modify
aZ:suggest
Z:modify
abZ:suggest
Z:modify
cZ:suggest
Z:clear

zsh: command not found: abc
Ted%
```

Notes:
1. I type `a` and get `Z:modify<enter>a`, and after 4 seconds `Z:suggest`
2. I type `b` and get  `Z:modify<enter>ab`, and after 4 seconds `Z:suggest`: Why is the `a` repeated here?
3. I type `c` and get `Z:modify<enter>c`, and after 4 seconds `Z:suggest`
4. I type enter and get `Z:clear<enter>`

**Comment by stanica** [comment]
Any updates on this? I'm getting the garbled characters when using Windows Terminal and the built-in terminal to VS Code. Git Bash seems to work fine, but only in the standalone terminal. Git Bash In VS Code is broken. Using v0.6.4 fixes the garbled text but the cursor briefly jumps back a character every time I enter a new character that's part of an autocompleted command.

**Comment by CervEdin** [comment]
I'm pretty sure this is a bug in the windows terminal and how it treats ANSI escape codes.
There are others similar reports like [this openjdk](https://bugs.openjdk.java.net/browse/JDK-8258808) and [this coc](https://github.com/neoclide/coc.nvim/issues/2309).
You can try and enable legacy console to and see if it resolves the issue

**Comment by dbachran** [comment]
> Oh, but I noted the following: I am using a custom console on Windows (https://github.com/cbucher/console), where I am seeing this issue. When trying to reproduce this issue with the Cygwin Terminal (mintty.exe) I do not get those funny characters. So there seems to be something different with regards to how these terminals/consoles behave?

I haven't tried Windows terminal, yet. (Actually, I don't even have it installed.)

All I want to say is: There are other consoles affected, as well.

**Comment by CervEdin** [comment]
I'm wondering if it has to do with the windows code page setting. I have a different machine with similar configuration where I don't see this issue. It has the code-page set to `utf-8`

![image](https://user-images.githubusercontent.com/3063442/145975350-ead21b7e-cb61-435d-a117-475406f393ea.png)

**Comment by CervEdin** [comment]
I was able to [solve the seemingly related issue](https://github.com/neoclide/coc.nvim/issues/2309#issuecomment-994559247) by

> changing the windows code page to `utf-8` [as per](https://superuser.com/a/1435645) resolved the issue on my machine

**Comment by ninagrosse** [comment]
I'm having the same problem with Git Bash, zsh and oh-my-zsh on Windows 10. However, not all terminals seem to be affected. I'm mostly using Tabby (https://github.com/Eugeny/tabby) and the integrated terminal of VSCode. 

When launching Git Bash via Tabby, it uses `$TERM=cygwin` which doesn't cause any issues. When I change it to `$TERM=xterm-256color`, the funny characters appear when entering a command with no availabe suggestions. VSCode terminal seems to use `$TERM=xterm-256color` by default, so same problem here. However, the default terminal of Git Bash (MINGW64 / MSYS2) doesn't have this problem when setting it to `xterm-256color`.

The workaround with disabling async mode via `unset ZSH_AUTOSUGGEST_USE_ASYNC` fixes the problem when using `xterm-256color` in Tabby and VSCode.

**Comment by ngoc199** [comment]
> You can use this:
> 
> ```
> cd ${ZSH_CUSTOM:-~/.oh-my-zsh/custom}/plugins/zsh-autosuggestions
> git checkout tags/v0.6.4 -b v0.6.4-branch
> ```

Since the changes made by @kevindqc work, could someone consider merging this to the current version?

**Comment by izancalderaro** [comment]
> You can use this:
> 
> ```
> cd ${ZSH_CUSTOM:-~/.oh-my-zsh/custom}/plugins/zsh-autosuggestions
> git checkout tags/v0.6.4 -b v0.6.4-branch
> ```

Works to me, thx!

**Comment by mbwhite** [comment]
FYI I've exactly this issue and I didn't have zsh-autosuggestions installed;  this is using windows terminal.. So I think the issue is more the terminal.

**Comment by CatBraaain** [comment]
> I just updated git for Windows and the problem is gone.
https://github.com/ohmyzsh/ohmyzsh/issues/12376#issuecomment-2088479110

I follow this comment and it worked.

Windows 10
Zsh version: 5.9
Git for Windows: 2.49.0
